docs/advanced/UsageWithTypescript.md

```diff
@@ -0,0 +1,219 @@
+
+## Type Checking State
+
+Type checking each slice of state is simple when you understand that each slice of state is just an object.
+
+Describing the state shape of the chat reducer's slice of state:
+
+```ts
+// src/store/chat/types.ts
+
+export interface IMessage {
+  user: string
+  message: string
+  timestamp: number
+}
+
+export interface IChatState {
+  messages: IMessage[]
+}
+```
+
+Describing the state shape of the system reducer's slice of state:
+
+```ts
+// src/store/system/types.ts
+
+export interface ISystemState {
+  loggedIn: boolean
+  session: string
+  userName: string
+}
+```
+
+Note that we are exporting these interfaces to reuse them later in application state and action creators.
+
+We can now create an interface which describes the global store:
+
+```ts
+// src/store/index.ts
+
+export interface IAppState {
+  system: ISystemState
+  chat: IChatState
+}
+```
+
+## Type Checking Actions & Action Creators
+
+We will be using TypeScript's enums to declare our action constants. [Enums](https://www.typescriptlang.org/docs/handbook/enums.html) allow us to define a set of named constants.
+
+Chat Action Constants and Actions:
+
+```ts
+// src/store/chat/types.ts
+
+export enum ChatActions {
+  SendMessage = '@CHAT:SEND/MESSAGE'
+}
+
+export interface ISendMessageAction {
+  type: ChatActions.SendMessage
+  payload: IMessage
+}
+```
+
+With these types we can now also type check chat's action creators:
+
+```ts
+// src/store/chat/actions.ts
+
+import { IMessage, ISendMessageAction, ChatActions } from './types'
+
+export type sendMessageType = (newMessage: IMessage) => ISendMessageAction
+
+export const sendMessage: sendMessageType = newMessage => ({
+  type: ChatActions.SendMessage,
+  payload: newMessage
+})
+```
+
+System Action Constants and Actions:
+
+```ts
+// src/store/system/types.ts
+
+export enum SystemActions {
+  UpdateSession = '@SYSTEM:UPDATE/SESSION'
+}
+
+export interface IUpdateSessionAction {
+  type: SystemActions.UpdateSession
+  payload: ISystemState
+}
+```
+
+With these types we can now also type check system's action creators:
+
+```ts
+// src/store/system/actions.ts
+
+import { SystemActions, IUpdateSessionAction, ISystemState } from './types'
+
+export type updateSessionType = (
+  newSession: ISystemState
+) => IUpdateSessionAction
+
+export const updateSession: updateSessionType = newSession => ({
+  type: SystemActions.UpdateSession,
+  payload: newSession
+})
+```
+
+## Type Checking Reducers
+
+Reducers are just pure functions that take the previous state, an action and then return the next state. Redux provides a `Reducer` type which can be imported to assist with type checking.
+
+Type checked chat reducer:
+
+```ts
+// src/store/chat/reducers.ts
+
+import { Reducer } from 'redux'
+import { IChatState, ChatActions } from './types'
+
+const initialState: IChatState = {
+  messages: []
+}
+
+export const chatReducer: Reducer<IChatState> = (
```

Comment:
I don't think the explicit `Reducer` type is necessary. TypeScripts strength comes in using as few explicit types as possible - Consider using this instead:

```ts
interface SendMessageAction {
   type: typeof ChatActions.SendMessage,
   payload: Message
}
function chatReducer(state: IChatState, action: SendMessageAction) {
  switch (action.type) {
    return {
      messages: [...state.messages, action.payload]
    }
  }
}
```

one downside of this approach is that while the type checking at compile time *is exhaustive*, your function *may receive* inputs that do not match the appropriate actions at runtime. For example, `{ type: '@@redux/init' }`. I don't think there's a way to work around this, though.